bigdecimal/src/lib.rs

```diff
@@ -0,0 +1,637 @@
+                // forward to val-ref
+                $imp::$method(self, &other)
+            }
+        }
+    }
+}
+
+macro_rules! forward_ref_val_binop {
+    (impl $imp:ident for $res:ty, $method:ident) => {
+        impl<'a> $imp<$res> for &'a $res {
+            type Output = $res;
+
+            #[inline]
+            fn $method(self, other: $res) -> $res {
+                // forward to ref-ref
+                $imp::$method(self, &other)
+            }
+        }
+    }
+}
+
+macro_rules! forward_val_ref_binop {
+    (impl $imp:ident for $res:ty, $method:ident) => {
+        impl<'a> $imp<&'a $res> for $res {
+            type Output = $res;
+
+            #[inline]
+            fn $method(self, other: &$res) -> $res {
+                // forward to ref-ref
+                $imp::$method(&self, other)
+            }
+        }
+    }
+}
+
+macro_rules! forward_ref_ref_binop {
+    (impl $imp:ident for $res:ty, $method:ident) => {
+        impl<'a, 'b> $imp<&'b $res> for &'a $res {
+            type Output = $res;
+
+            #[inline]
+            fn $method(self, other: &$res) -> $res {
+                // forward to val-ref
+                $imp::$method(self.clone(), other)
+            }
+        }
+    }
+}
+
+// Forward everything to ref-ref, when reusing storage is not helpful
+macro_rules! forward_all_binop_to_ref_ref {
+    (impl $imp:ident for $res:ty, $method:ident) => {
+        forward_val_val_binop!(impl $imp for $res, $method);
+        forward_val_ref_binop!(impl $imp for $res, $method);
+        forward_ref_val_binop!(impl $imp for $res, $method);
+    };
+}
+
+/// A big decimal type.
+///
+#[derive(Clone, Debug, Hash)]
+#[cfg_attr(feature = "rustc-serialize", derive(RustcEncodable, RustcDecodable))]
+pub struct BigDecimal {
+    int_val: BigInt,
+    scale: i64,
+}
+
+impl BigDecimal {
+    /// Creates and initializes a `BigDecimal`.
+    ///
+    #[inline]
+    pub fn new(digits: BigInt, scale: i64) -> BigDecimal {
+        BigDecimal {
+            int_val: digits,
+            scale: scale,
+        }
+    }
+
+    /// Creates and initializes a `BigDecimal`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// // assert_eq!(BigDecimal::parse_bytes(b"0", 16), BigDecimal::zero());
+    /// // assert_eq!(BigDecimal::parse_bytes(b"f", 16), BigDecimal::parse_bytes(b"16", 10));
+    /// ```
+    #[inline]
+    pub fn parse_bytes(buf: &[u8], radix: u32) -> Option<BigDecimal> {
+        str::from_utf8(buf).ok().and_then(|s| BigDecimal::from_str_radix(s, radix).ok())
+    }
+
+    pub fn set_scale(&self, new_scale: i64) -> BigDecimal {
+        if self.scale == new_scale {
+            return self.clone();
+        }
+
+        if self.int_val.is_zero() {
+            return BigDecimal::new(BigInt::zero(), new_scale);
+        }
+
+        if new_scale > self.scale {
+            let raise = new_scale - self.scale;
+            // println!("raise: {}", raise);
+            let exp = (10 as i64).pow(raise as u32);
```

Comment:
This has been fixed by looping and multiplying BigInts until correct number is reached. I don't know if there's a better way of doing this. The only other way I could think of is formatting the string `BigDecimal::from_str(&format!("1e{}", raise));` which does pass my tests and probably has less room for error, but for some reason, it seems wrong.

I think it'd be nice to have a BigInt (or BigUint) `::as_order_of_magnitude(u64) -> Self` method that would handle this in the best way for us. Does something like this exist?